lib/connection_test.ts

```diff
@@ -0,0 +1,108 @@
+  postMessageFromIframeToSelf,
+} from "./shared/testing/messaging";
+
+describe("awaitConnectionFromIframe", () => {
+  cleanDomAfterEach();
+  beforeAll(addMessagePortMatchers);
+
+  it("should perform the handshake and transfer a connected port", async () => {
+    const iframe = document.createElement("iframe");
+    document.body.appendChild(iframe);
+    const portPromise = awaitConnectionFromIframe(iframe, origin);
+    const { port1, port2 } = new MessageChannel();
+    postMessageFromIframeToSelf(iframe, { [VERSION_KEY]: VERSION }, [port1]);
+    await expectAsync(await portPromise).toBeEntangledWith(port2);
+  });
+
+  it("should ignore messages from other windows", async () => {
+    const iframe = document.createElement("iframe");
+    document.body.appendChild(iframe);
+    const otherIframe = document.createElement("iframe");
+    document.body.appendChild(otherIframe);
+    const portPromise = awaitConnectionFromIframe(iframe, origin);
+    postMessageFromIframeToSelf(otherIframe, "wrong payload", []);
+    await expectAsync(portPromise).toBePending();
+    postMessageFromIframeToSelf(iframe, { [VERSION_KEY]: VERSION }, [
+      new MessageChannel().port1,
+    ]);
+    await portPromise;
+  });
+
+  it("should reject on origin mismatch", async () => {
+    const iframe = document.createElement("iframe");
+    document.body.appendChild(iframe);
+    const portPromise = awaitConnectionFromIframe(
+      iframe,
+      "https://wrong-origin.example"
+    );
+    postMessageFromIframeToSelf(iframe, { [VERSION_KEY]: VERSION }, [
+      new MessageChannel().port1,
+    ]);
+    await expectAsync(portPromise).toBeRejectedWithError();
+  });
+
+  it("should reject on nullish handshake message", async () => {
+    const iframe = document.createElement("iframe");
+    document.body.appendChild(iframe);
+    const portPromise = awaitConnectionFromIframe(iframe, origin);
+    postMessageFromIframeToSelf(iframe, null, [new MessageChannel().port1]);
+    await expectAsync(portPromise).toBeRejectedWithError();
+  });
+
+  it("should reject on wrong version", async () => {
+    const iframe = document.createElement("iframe");
+    document.body.appendChild(iframe);
+    const portPromise = awaitConnectionFromIframe(iframe, origin);
+    postMessageFromIframeToSelf(iframe, { [VERSION_KEY]: "wrong version" }, [
+      new MessageChannel().port1,
+    ]);
+    await expectAsync(portPromise).toBeRejectedWithError();
+  });
+
+  it("should reject on wrong number of ports", async () => {
+    const iframe = document.createElement("iframe");
+    document.body.appendChild(iframe);
+    const portPromise = awaitConnectionFromIframe(iframe, origin);
+    postMessageFromIframeToSelf(iframe, { [VERSION_KEY]: VERSION }, []);
+    await expectAsync(portPromise).toBeRejectedWithError();
+  });
+
+  const POSTMESSAGE_SHAREDARRAYBUFFER_SRCDOC =
+    "<!DOCTYPE html><title>Helper</title><script>parent.postMessage(new SharedArrayBuffer(), '*');</script>";
```

Comment:
Can we do this without SharedArrayBuffer?  Chrome 91 is restricting it to COEP pages.

Reply:
I am not aware of any way to do this, but I can make inquiries with Karma folks and see what needs to happen in order for the test runner to be using COEP. I think various teams internally will be having this problem, so hopefully there's some kind of solution in the works.

Reply:
I don't think we should make the test runner use COEP, since publisher pages will (mostly) not use COEP

Can we implement without SharedArrayBuffer?

Reply:
I had thought this wasn't possible but checked Web Platform Tests and found another way. Sent #9.